### PR TITLE
combine color and opacity in shaders

### LINKED ISSUE #1331: fix combining color and opacity in data driven styles
Color and opacity aren't correctly combined if either of them are data driven.

Either we need to:
A.  always do the color and opacity multiplication in the vertex shader
B. implement support for zoom and data driven properties

We should do **A** because it's easier.

### PR DESCRIPTION
**Pull request by ansis**
Fixes #1331 

@lucaswoj can you do a quick review?

Some render tests are now failing:
- background-opacity color
- fill-color opacity
- fill-opacity function
- line-pattern opacity

All of them look identical but have tiny differences in every pixel. Multiplying opacity by 255 in js and the shader probably have slightly different results. @kkaefer does that sound right, or do you think there's a small bug somewhere?

### CODE REVIEW

File: js/render/draw_circle.js [hunk 1]
```diff
@@ -63,9 +70,8 @@ function drawCircles(painter, layer, posMatrix, tile) {
                 layer.layout[property.styleName]
             );
 
-            // TODO remove this via https://github.com/mapbox/mapbox-gl-js/issues/1319
-            if (property.styleName === 'circle-color') {
-                value = [value[0] * 255, value[1] * 255, value[2] * 255, value[3] * 255];
```
Comment (lucaswoj) [inline js/render/draw_circle.js]: Why did you delete these lines?

**Comment by lucaswoj** [comment]
This looks :+1: overall. 

Did you move the `* 255` for all the RGBA vales somewhere? I see that you deleted all the multiplication operations in the `draw_*` methods but don't see to where you moved the operation. Related to #1319

**Comment by ansis** [comment]
I moved them into `premultiplyLayer` in `style/style_layer.js`

**Comment by lucaswoj** [comment]
@ansis Cool. Lets wait to hear from @kkaefer before :ship: but I'm excited about this!

**Comment by jfirebaugh** [comment]
@ansis Is there a particular reason why you multiply opacity values by 255 in js rather than keeping them in `[0, 1]` all the way to the shader?

File: js/style/style_layer.js [hunk 2]
```diff
@@ -178,20 +178,15 @@ function premultiplyLayer(layer, type) {
...
+        outlineColor = layer[outlineProp];
 
-    var colorOpacity = color && (opacity * color[3]);
-    var haloOpacity = haloColor && (opacity * haloColor[3]);
-    var outlineOpacity = outlineColor && (opacity * outlineColor[3]);
-
-    if (colorOpacity !== undefined && colorOpacity < 1) {
-        layer[colorProp] = util.premultiply([color[0], color[1], color[2], colorOpacity]);
+    if (color) {
+        layer[colorProp] = util.premultiply([color[0], color[1], color[2], color[3] * 255]);
```
Comment (jfirebaugh) [inline js/style/style_layer.js]: Is `util.premultiply([color[0], color[1], color[2], color[3] * 255])` correct? It seems like `util.premultiply` expects a 0-1 opacity value.

**Comment by ansis** [comment]
@jfirebaugh when we store colours in the vertex buffers for data driven styles each component of a colour is stored in an unsigned byte. Since we need to divide by 255 in the shaders for data driven values we need to convert constant/zoom-based values to 0-255 as well.

File: js/style/style_layer.js (hunk 2 above)
Comment (ansis) [inline js/style/style_layer.js]: hmm, maybe this isn't the clearest place to do the multiplication by 255. The result is correct because multiplying the alpha value by 255 before premultiplication is the same as multiplying all components by 255 after premultiplication. I can move it elsewhere tomorrow.

**Comment by kkaefer** [comment]
Did you try using `(a_color / 255.0) * (a_opacity / 255.0)`?

**Comment by ansis** [comment]
> Did you try using (a_color / 255.0) * (a_opacity / 255.0)?

I tried just now and it is the same.
